Added to_gate method on certain operators

shifubear opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
This pull request resolves issue #1091. I implemented a to_gate method for the following operators: 
- MatrixOp
- CircuitOp
- PauliOp
- EvolvedOp

### Details and comments
I added the operator to a circuit and created a gate from that circuit. Is this the correct way to go about it moving forward? 
I looked through the other operators, but I couldn't find any others that had the to_instruction method implemented. If there are any more that need this method, I'd be happy to implement this on those. 
I added tests for MatrixOp, but there weren't existing tests for the to_instruction method for the other three operators. I was wondering if these need some extra tests, in addition to a test for this added method.

shifubear commented:
So it seems like the error I'm getting from the latest version of pylint is a known bug: https://github.com/PyCQA/pylint/issues/3537
where pylint doesn't recognize the 'typing' types as a valid type. The following is the warning I'm getting. 
`qiskit/aqua/operators/primitive_ops/matrix_op.py:223:11: W1116: Second argument of isinstance is not a type (isinstance-second-argument-not-valid-type)`

Can I just ignore this warning and assume everything is working correctly?

Cryoris commented on qiskit/aqua/operators/primitive_ops/circuit_op.py:
```diff
@@ -212,6 +212,10 @@ def to_circuit_op(self) -> OperatorBase:
     def to_instruction(self) -> Instruction:
         return self.primitive.to_instruction()  # type: ignore
 
+    # def to_gate(self) -> Gate:
```
Why is this commented out? 🙂

shifubear replied:
I just added skeleton comments to the operators that I think will need this method for later on 😅

Cryoris commented on qiskit/aqua/operators/evolutions/evolved_op.py:
```diff
@@ -159,3 +159,7 @@ def log_i(self, massive: bool = False) -> OperatorBase:
     def to_instruction(self, massive: bool = False) -> Instruction:
         return self.primitive.to_matrix_op(massive=massive).exp_i()  # type: ignore
+
+    # def to_gate(self, massive: bool = False) -> Gate:
```
This should use the `HamiltonianGate`, see the `to_circuit` (or `to_circuit_op`)

shifubear replied:
Noted!

Cryoris commented on qiskit/aqua/operators/primitive_ops/matrix_op.py:
```diff
@@ -218,5 +218,15 @@ def to_matrix_op(self, massive: bool = False) -> OperatorBase:
 
+    def to_gate(self, parameter_map=None, label=None) -> Gate:
+        """ Returns a ``Gate`` equivalent to this Operator. """
+        if isinstance(self.primitive, List):  # pylint: disable=typecheck
```
Can you explain what you did here? Shouldn't the primitive always be a matrix? Then you could return a `UnitaryGate` and pass the matrix as input.

shifubear replied:
Ohh that's a much better way to do it. I'll try it using that method! Thanks 😄

Cryoris commented on qiskit/aqua/operators/primitive_ops/pauli_op.py:
```diff
@@ -306,6 +306,10 @@ def to_instruction(self) -> Instruction:
 
         return self.to_circuit().to_instruction()
 
+    # def to_gate(self) -> Gate:
```
Why is this commented?

shifubear replied:
I rewrote this method so it's not commented out anymore!

woodsp-ibm commented:
> So it seems like the error I'm getting from the latest version of pylint is a known bug: [PyCQA/pylint#3537](https://github.com/PyCQA/pylint/issues/3537)
> where pylint doesn't recognize the 'typing' types as a valid type. The following is the warning I'm getting.
> `qiskit/aqua/operators/primitive_ops/matrix_op.py:223:11: W1116: Second argument of isinstance is not a type (isinstance-second-argument-not-valid-type)`
> 
> Can I just ignore this warning and assume everything is working correctly?

That is this line below right...
```
        if isinstance(self.primitive, List):  # pylint: disable=typecheck
```
use the Python `list` type not the Typehint `List`
```
        if isinstance(self.primitive, list): 
```

shifubear commented:
I've fixed the sections of concern highlighted by @Cryoris and @woodsp-ibm and also added the to_gate operator on the three other gates. Are there any other gates that need this operator that I might have missed?

woodsp-ibm commented on test/aqua/operators/test_op_construction.py:
```diff
@@ -222,6 +247,23 @@ def test_matrix_to_instruction(self):
+        """Test MatrixOp.to_gate yields a Gate object."""
+        matop = (H ^ 3).to_matrix_op()
+        with self.subTest('assert to_gate returns Gate'):
+            self.assertIsInstance(matop.to_gate(), Gate)
```
Is there some check we can do on the gate to make sure that the conversion code in to_gate() is working beyond just checking its type. Can we check the value in some way - for instance doing to_matrix on the gate and assuring this is as expected too, or whatever seems reasonable to check that the gate is what we expect beyond just checking type.

woodsp-ibm commented:
@Cryoris This I think, if its still wanted needs to be done on the ported opflow now in terra repo but closed off here. Do you have an equivalent issue to #1091 in Terra if this is still wanted there?

woodsp-ibm commented:
This has not been worked on in some time and needs work to complete it. Since the opflow is now in Terra I am moving the issue over there, to be potentially addressed at some point, and closing this PR. @shifubear if you are still interested in doing this and completing the work it will need to be completed over in the opflow in Terra.
